### DISCUSSION TITLE
replacing default discover to enable dynamic test filter

### DISCUSSION THREAD

**Discussion by livioc** [xunit/xunit#2226] [Question] [answered]
hi, I'm using the latest nuget version of xunit for the .net framework. I'd like to add a TargetEnvironment trait on each test that represents an UAT test for one+ specific environments (for example [Trait("TargetEnvironment", Environment.Dev | Environment.Production)], that is because I have some tests that must be run only on some specific environments.
then I want to use a dynamic setting value (for example loaded from the appSettings section) that configures the current environment (Dev, Test, Production, etc) and a custom test discover that will load only the tests without a TargetEnvironment trait or having it with a value that contains the current env value.
My question is about a way to replace the default test discover Xunit.Sdk.FactDiscoverer of the FactAttribute or other suggestions, at the moment the only solution I found is to replace all the FactAttribute with a custom subclass

**Comment by bradwilson** [comment, marked as answer]
In v3, we've enabled in-box dynamic skipping (you can do that today in v2 with a [third-party extension](https://www.nuget.org/packages/Xunit.SkippableFact/)).

Note that this is slightly different behavior than what you're asking for, which is: the test still exists, but shows up as a skipped (you would presumably provide a reason for skipping which describes an incompatible environment).

As for replacing the default `FactDiscoverer`, there's no way to do that today in v2 without major surgery. If you open an issue for that, I'll put it on the v3 backlog.

**Comment by livioc** [comment]
Hi, thanks for your reply, the dynamic skipping doesn't work very well in my scenario because I want to configure the allowed environments using some attribute on the test method and for use the dynamic skipping I need to use some complex runtime reflection code to get their value from the test method body.
eventually I'll use a custom FactAttribute with a custom FactDiscover, I don't think that the benefit of allowing the replacing of the default FactDiscover outweights the large refactoring work for it if it's a so major surgery.
thanks again :)
